src/Altinn.App.Core/Features/Maskinporten/MaskinportenClient.cs

```diff
@@ -0,0 +1,272 @@
+using System.Text.Json;
+using Altinn.App.Core.Features.Maskinporten.Exceptions;
+using Altinn.App.Core.Features.Maskinporten.Models;
+using Microsoft.Extensions.Caching.Hybrid;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.IdentityModel.JsonWebTokens;
+using Microsoft.IdentityModel.Tokens;
+
+namespace Altinn.App.Core.Features.Maskinporten;
+
+/// <inheritdoc/>
+public sealed class MaskinportenClient : IMaskinportenClient
+{
+    /// <summary>
+    /// The margin to take into consideration when determining if a token has expired (seconds).
+    /// <remarks>This value represents the worst-case latency scenario for <em>outbound</em> connections carrying the access token.</remarks>
+    /// </summary>
+    internal const int TokenExpirationMargin = 30;
+
+    private static HybridCacheEntryOptions _defaultCacheExpiration = CacheExpiry(TimeSpan.FromSeconds(60));
+
+    private readonly ILogger<MaskinportenClient> _logger;
+    private readonly IOptionsMonitor<MaskinportenSettings> _options;
+    private readonly IHttpClientFactory _httpClientFactory;
+    private readonly TimeProvider _timeprovider;
+    private readonly HybridCache _tokenCache;
+    private readonly Telemetry? _telemetry;
+
+    /// <summary>
+    /// Instantiates a new <see cref="MaskinportenClient"/> object.
+    /// </summary>
+    /// <param name="options">Maskinporten settings.</param>
+    /// <param name="httpClientFactory">HttpClient factory.</param>
+    /// <param name="tokenCache">Token cache store.</param>
+    /// <param name="logger">Logger interface.</param>
+    /// <param name="timeProvider">Optional TimeProvider implementation.</param>
+    /// <param name="telemetry">Optional telemetry service.</param>
+    public MaskinportenClient(
+        IOptionsMonitor<MaskinportenSettings> options,
+        IHttpClientFactory httpClientFactory,
+        HybridCache tokenCache,
+        ILogger<MaskinportenClient> logger,
+        TimeProvider? timeProvider = null,
+        Telemetry? telemetry = null
+    )
+    {
+        _options = options;
+        _telemetry = telemetry;
+        _timeprovider = timeProvider ?? TimeProvider.System;
+        _logger = logger;
+        _httpClientFactory = httpClientFactory;
+        _tokenCache = tokenCache;
+    }
+
+    /// <inheritdoc/>
+    public async Task<MaskinportenTokenResponse> GetAccessToken(
+        IEnumerable<string> scopes,
+        CancellationToken cancellationToken = default
+    )
+    {
+        string formattedScopes = FormattedScopes(scopes);
+        DateTimeOffset referenceTime = _timeprovider.GetUtcNow();
+
+        _telemetry?.StartGetAccessTokenActivity(_options.CurrentValue.ClientId, formattedScopes);
+
+        var result = await _tokenCache.GetOrCreateAsync(
+            formattedScopes,
```

Comment:
Er det garantert at `formattedScopes` er unikt og ikke vil gi en navnekollisjon nå som du bruker er delt cache?

Reply:
Forsåvidt ingen garanti for det, når jeg tenker meg om. Kan padde litt med en unik identifier kanskje?

Reply:
Ja, det er nok lurt. Utrolig vanskelig å feilsøke cache key kollisjoner.

Reply:
Jeg elsker salt både i maten og i koden, så det her blir bra